rasterio/_err.pyx

```diff
@@ -205,11 +215,40 @@ cdef OGRErr exc_wrap_ogrerr(OGRErr err) except -1:
         raise CPLE_BaseError(3, err, "OGR Error code {}".format(err))
 
 
+@contextlib.contextmanager
+def stack_errors():
+    """TODO: better name."""
+    # Set up.
+    CPLErrorReset()
+    error_stack = []
+
+    # chaining_error_handler (better name a TODO) records GDAL errors
+    # in the order they occur and converts to exceptions.
+    CPLPushErrorHandlerEx(chaining_error_handler, <void *>error_stack)
+
+    # Run code in the `with` block.
+    yield
```

Comment:
Inside the `with` block `chaining_error_handler` can be appending objects to `error_stack`.